homeworks/alvaxgln/5-remesh/src/geometry/halfedge.ts

```diff
@@ -0,0 +1,504 @@
+
+        // point the new halfedge and twin to each other
+        bhe.twin = <Halfedge>current;
+        current!.twin = bhe;
+
+        current = next;
+      } while (current !== he);
+
+      // link the cycle of boundary halfedges together
+      const n = bcycle.length;
+      for (let j = 0; j < n; j++) {
+        bcycle[j].next = bcycle[(j + n - 1) % n];
+        bcycle[j].prev = bcycle[(j + 1) % n];
+        hasTwin.set(bcycle[j], true);
+        hasTwin.set(bcycle[j].twin, true);
+      }
+    }
+    this.resetIndices();
+  }
+  /**
+   * resetIndices clears all null elements and reset the indices
+   */
+  resetIndices() {
+    // clear null elements
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const notnull = (e: any) => (e === null ? false : true);
+    this.verts = this.verts.filter(notnull);
+    this.edges = this.edges.filter(notnull);
+    this.faces = this.faces.filter(notnull);
+    this.halfedges = this.halfedges.filter(notnull);
+    this.boundaries = this.boundaries.filter(notnull);
+
+    // reset indices
+    let index = 0;
+    this.verts.forEach(v => {
+      v!.idx = index++;
+    });
+    index = 0;
+    this.edges.forEach(e => {
+      e!.idx = index++;
+    });
+    index = 0;
+    this.faces.forEach(f => {
+      f!.idx = index++;
+    });
+    index = 0;
+    this.halfedges.forEach(h => {
+      h!.idx = index++;
+    });
+    index = 0;
+    this.boundaries.forEach(b => {
+      b.idx = index++;
+    });
+  }
+
+  simplify(reduceRatio: number) {
+    // TODO: implement QEM simplification algorithm
+    //
+    // Hints:
+    //    - Boundary edges maybe skipped without handling
+    //    - Use the provided EdgeQueue to maintain a priority queue
+    //    - If necessary, use the provided resetIndices
+    //    - Carefully maintain the edge collapse as it may easily be wrong
+    //      It is recommended to first start with a simply cube object.
+    //    - To mark remove an object, one can assign null to the array, then
+    //      calling resetIndices will filter out all null elements. (Only call
+    //      at the end of each simplification round)
+    //
+    // Additional Task:
+    //    - Be able to handle meshes with boundary (need preserve the boundary)
+
+    // Number of faces that should be removed.
+    const reducedFaces = Math.floor(this.faces.length * reduceRatio);
+
+    //create edgequeue
+    let equeue = new EdgeQueue( (edge, edge2)=>{
+      return edge.error() < edge2.error();
+    })
+
+    //enqueue edges
+    this.edges.forEach(e=>{
+      if (e!=null){
+        equeue.push(e);
+      }
+    })
+
+
+    //console.log("Test 1")
+
+    //to test
+    //const num_faces = this.faces.length -22000;
+
+    //Amount of faces remaining after reduction
+    const num_faces = this.faces.length-reducedFaces;
+
+
+     while(this.faces.length > num_faces){
+
+      equeue = new EdgeQueue( (edge, edge2)=>{
+        return edge.error() < edge2.error();
+      })
```

Comment:
Reconstructing the queue seems to be the root reason that the program is slow.
Maintaining a single queue will significantly speedup the process.